csharp_dotnetcore/17.Proactive-Messages/ProactiveBot.cs

```diff
@@ -0,0 +1,208 @@
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // See https://aka.ms/about-bot-activity-message to learn more about the message and other activity types.
+            if (turnContext.Activity.Type != ActivityTypes.Message)
+            {
+                // Handle non-message activities.
+                await OnSystemActivityAsync(turnContext);
+            }
+            else
+            {
+                // Get the job log from the turn context.
+                JobLog jobLog = await StateAccessors.JobLogData.GetAsync(turnContext, () => new JobLog());
+
+                // Get the user's text input for the message.
+                var text = turnContext.Activity.Text.Trim().ToLowerInvariant();
+                switch (text)
+                {
+                    case "run":
+                    case "run job":
+
+                        // Start a virtual job for the user.
+                        JobLog.JobData job = CreateJob(turnContext, jobLog);
+                        ConversationReference conversation = turnContext.Activity.GetConversationReference();
+
+                        await turnContext.SendActivityAsync(
+                            $"We're starting job {job.TimeStamp} for you. We'll notify you when it's complete.");
+
+                        break;
+
+                    case "show":
+                    case "show jobs":
+
+                        // Display information for all jobs in the log.
+                        if (jobLog.Count > 0)
+                        {
+                            await turnContext.SendActivityAsync(
+                                "| Job number &nbsp; | Conversation ID &nbsp; | Completed |<br>" +
+                                "| :--- | :---: | :---: |<br>" +
+                                string.Join("<br>", jobLog.Values.Select(j =>
+                                    $"| {j.TimeStamp} &nbsp; | {j.Conversation.Conversation.Id} &nbsp; | {j.Completed} |")));
+                        }
+                        else
+                        {
+                            await turnContext.SendActivityAsync("The job log is empty.");
+                        }
+
+                        break;
+
+                    default:
+
+                        // Check whether this is simulating a job completed event.
+                        string[] parts = text?.Split(' ', StringSplitOptions.RemoveEmptyEntries);
+                        if (parts != null && parts.Length is 2
+                            && parts[0].Equals("done", StringComparison.InvariantCultureIgnoreCase)
```

Comment:
I've... never seen this before. Ever. The closes is in OData where they write ("if x eq 2") to get around the URL encoding issues. 

"parts.Length is 2" ? What does that do?

Reply:
Interesting, no idea  Changing to '=='.  I'll ping Jonathan to make sure it's not something subtle here.

Reply:
It's just some C# syntactic sugar for [pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/is#pattern-matching-with-is).